A43-sdk-authorization.md

```diff
@@ -0,0 +1,713 @@
+A43: SDK authorization
+----
+* Author(s): [Ashitha Santhosh](https://github.com/ashithasantosh)
+* Approver: [Doug Fawley](https://github.com/dfawley),
+            [Eric Anderson](https://github.com/ejona86),
+            [Mark Roth](https://github.com/markdroth),
+            [Sanjay Pujare](https://github.com/sanjaypujare),
+            [Yihua Zhang](https://github.com/yihuazhang)
+* Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2021-08-30
+* Discussion at: https://groups.google.com/g/grpc-io/c/m4Lns0tO-fE
+
+## Abstract
+
+gRPC is currently implementing an open-source standard authorization solution.
+This proposal discusses how this framework would be made available to OSS gRPC
+users (like network switches) to perform per-RPC authorization checks.
+
+## Background
+
+Authentication verifies the identity of the requestor. Authorization ensures that
+the requestor has sufficient access to make a particular request.
+
+gRPC offers built-in authentication mechanisms like TLS, ALTS and even allow
+users to plug in their own authentication systems. However, gRPC doesn't have any
+standard authorization solution available. If a user wanted to perform per RPC
+authorization check, they have to implement their own solution in the application.
+The standard gRPC authorization mechanism described in this proposal allows users
+to simply provide an authorization policy for their server. Then, for each RPC,
+the server verifies whether the client is authorized to issue the RPC.
+
+## Proposal
+
+### User facing Authorization Policy
+
+gRPC SDK authorization policy is the user facing policy language that allows
+service owners and/or security admins to enable per-RPC authorization checks.
+
+SDK authorization policy is extendable, we may add new fields in the future as
+new requirements are introduced, we won’t be removing fields or updating
+semantics of existing fields. However, this means if user wants to use policy
+with new fields, they need to use the right gRPC version. Using policy with new
+fields, with an older version of gRPC, will result in policy being marked invalid.
+This is essential, because otherwise we may end up allowing an RPC request without
```

Comment:
If a policy's deny rule (but not allow rules) uses new fields and you mark the policy invalid you will end up allowing an RPC that could otherwise have been denied. In such a case your statement doesn't apply and you are not mitigating a security risk. Isn't that right?

Reply:
We will be returning error status back to application notifying them of unknown/new fields. Previously, if we were using an older version, and deny rules had some new fields, then we would have just ignored that field, and evaluated based on the other fields (which are supported by that version), so we could have ended up allowing a request that should have been denied. Here instead of silently ignoring those new fields, we are returning an error back to user. In case of FW, the errors will be logged (except first reload, where we return error). Then the user can choose to fix the version or update the policy.